### ISSUE TITLE
Two suggestions for more DRY code

### ISSUE THREAD

**Issue by leonardo-m** [Manishearth/rust-clippy#1674]
Given this code:

```
struct Foo { cap: usize }
impl Foo {
    fn new() -> Foo { Foo { cap: 1 } }
    fn with_capacity(cap: usize) -> Foo {
        Foo { cap: cap }
    }
}
fn main() {}
```

In my opinion Clippy should suggest to replace four of those usages of "Foo" with "Self" and to remove one ": cap", making the code more DRY, the intended resulting code should be:

```
struct Foo { cap: usize }
impl Foo {
    fn new() -> Self { Self { cap: 1 } }
    fn with_capacity(cap: usize) -> Self {
        Self { cap }
    }
}
fn main() {}
```

**Comment by shssoichiro** [comment]
I can't comment on the use of `Self`, but the use of shorthand property syntax is something that `rustfmt` does, so I don't think clippy needs to lint for it.

**Comment by oli-obk** [comment]
rustfmt can't do the Self transformation, as it needs some type info for that I think

**Comment by oli-obk** [comment]
We do have this lint now. I'm leaving the issue open to solve the false negatives still occurring:

```rust
struct Foo<'a>(PhantomData<&'a i32>);
impl<'a> Foo<'a> {
    // does not trigger the lint, because of lifetimes
    // lifetimes matching the impl block's lifetime should be fine though
    fn foo() -> Foo<'a> { unimplemented!() }
}
```

I also have some concerns relating generics which we should check:

```rust
struct Foo<T>(T);

impl<T> Foo<T> {
    fn bar<U>(u: U) -> Foo<U> { unimplemented!() } // should not lint
    fn baa(t: T) -> Foo<T> { unimplemented!() } // should lint
}
```
